src/lib/unlang/module.c

```diff
@@ -864,6 +871,85 @@ static void unlang_module_event_retry_handler(UNUSED fr_event_list_t *el, fr_tim
 	frame->signal(request, frame, FR_SIGNAL_TIMEOUT);
 }
 
+/** Parse the result of module_env tmpl expansion
+ */
+static int module_env_value_parse (request_t *request, unlang_frame_state_module_t *state) {
+	fr_value_box_t			*vb, *array;
+	void				*out;
+	module_env_parsed_t const	*env = state->last_expanded;
+	fr_type_t			type = FR_BASE_TYPE(env->rule->type);
+
+	if (type == FR_TYPE_SUBSECTION) {
+		return 0;
+	}
+
+	vb = fr_value_box_list_head(&state->tmpl_expanded);
+
+	if (!vb) {
+		if (env->rule->pair.required) {
+			RPEDEBUG("Failed to evaluate required module option %s", env->rule->name);
+			return -1;
+		}
+		return 0;
+	}
+
+	/*
+	 *	Concatenate multiple boxes if needed
+	 */
+	if (env->rule->pair.concat &&
+	    fr_value_box_list_concat_in_place(state->env_data, vb, &state->tmpl_expanded,
+					      type, FR_VALUE_BOX_LIST_FREE, true, SIZE_MAX) < 0 ) {
+		RPEDEBUG("Failed concatenating values for %s", env->rule->name);
+		return -1;
+	}
+
+	if (env->rule->pair.single && (fr_value_box_list_num_elements(&state->tmpl_expanded) > 1)) {
+		RPEDEBUG("%d values found for %s.  Only one is allowed",
+			 fr_value_box_list_num_elements(&state->tmpl_expanded), env->rule->name);
+		return -1;
+	}
+
+	/*
+	 *	Find the location of the output
+	 */
+	out = ((uint8_t *)state->env_data) + env->rule->offset;
+
+	/*
+	 *	If this is a multi pair option, the output is an array.
+	 *	On the first pass allocate the array.
+	 */
+	if (env->rule->pair.multi) {
+		if (env->multi_index == 0) {
+			array = talloc_zero_array(state->env_data, fr_value_box_t, env->opt_count);
+			*(void **)out = array;
+		} else {
+			array = *(fr_value_box_t **)out;
+		}
+		out = &array[env->multi_index];
+	}
+
+	while ((vb = fr_value_box_list_pop_head(&state->tmpl_expanded))) {
+		switch(type) {
+		case FR_TYPE_VOID:
+			break;
+
+		case FR_TYPE_LEAF:
+			if (fr_value_box_cast_in_place(state->env_data, vb, type, NULL) < 0) {
```

Comment:
Same here, should probably be the multi array if one's configured

Reply:
It seems a bit simpler to initialise the output array outside of this function before the tmpl is expanded, pass it as the ctx to the tmpl expansion code, and pass a pointer to the correct output element into this function.